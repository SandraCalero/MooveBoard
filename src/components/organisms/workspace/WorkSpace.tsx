import Header from 'components/molecules/Header/Header';
import PostIt from '../../atoms/PostIt/PostIt';

export default function WorkSpace() {
	return (
		<>
			<Header />
			<PostIt />
		</>
	);
}
