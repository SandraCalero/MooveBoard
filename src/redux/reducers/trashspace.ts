/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IPostIt, IPostItEditable } from 'globals/definitions/postItProps';

interface trashspaceStateProps {
	deletedPostIts: Array<IPostIt>;
}

const localDeletedPostIts = localStorage.getItem('deletedPostIts');

const deletedPostIts: Array<IPostIt> = localDeletedPostIts
	? JSON.parse(localDeletedPostIts)
	: [];

const initialState: trashspaceStateProps = {
	deletedPostIts,
};

const trashspaceSlice = createSlice({
	name: 'trashspace',
	initialState,
	reducers: {
		addPostItToTrash: (state, action: PayloadAction<IPostIt>) => {
			state.deletedPostIts.unshift({
				id: action.payload.id,
				content: action.payload.content,
			});
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		addAllToTrash: (state, action: PayloadAction<Array<IPostItEditable>>) => {
			const newDeletedPostIts = action.payload.map((postIt) => ({
				id: postIt.id,
				content: postIt.content,
			}));
			state.deletedPostIts = [...state.deletedPostIts, ...newDeletedPostIts];
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		clearTrashBin: (state) => {
			state.deletedPostIts = [];
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		deletePostIt: (state, action: PayloadAction<number>) => {
			const postItIndex = state.deletedPostIts.findIndex(
				(postIt) => postIt.id === action.payload
			);
			state.deletedPostIts.splice(postItIndex, 1);
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
	},
});

export const { addPostItToTrash, addAllToTrash, clearTrashBin, deletePostIt } =
	trashspaceSlice.actions;

export default trashspaceSlice.reducer;
