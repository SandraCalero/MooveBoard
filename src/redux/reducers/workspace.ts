/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IPostIt, IPostItEditable } from 'globals/definitions/postItProps';
import { sortAsc, sortDes } from 'utils/orderBy';

type PostItId = Pick<IPostIt, 'id'>;

interface workspaceStateProps {
	currentPostItId: number;
	postItList: Array<IPostItEditable>;
	orderBy: string;
	darkMode: boolean;
}

const localPostItList = localStorage.getItem('postItList');
const localCurrentPostItId = localStorage.getItem('currentPostItId');
const localDarkMode = localStorage.getItem('darkMode');
const localOrdeBy = localStorage.getItem('orderBy');

const postItList: Array<IPostItEditable> = localPostItList
	? JSON.parse(localPostItList)
	: [];

const currentPostItId: number = localCurrentPostItId
	? JSON.parse(localCurrentPostItId)
	: 0;

const darkMode: boolean = localDarkMode ? JSON.parse(localDarkMode) : false;

const orderBy: string = localOrdeBy || 'DES';

if (darkMode) {
	document.documentElement.classList.toggle('dark');
}

const initialState: workspaceStateProps = {
	currentPostItId,
	postItList,
	darkMode,
	orderBy,
};

const workspaceSlice = createSlice({
	name: 'workspace',
	initialState,
	reducers: {
		createPostIt: (state) => {
			state.currentPostItId += 1;
			state.postItList.unshift({
				id: state.currentPostItId,
				content: '',
				disabled: true,
			});
			if (state.orderBy === 'DES') {
				sortDes(state.postItList);
			} else {
				sortAsc(state.postItList);
			}
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
			localStorage.setItem(
				'currentPostItId',
				JSON.stringify(state.currentPostItId)
			);
		},
		restorePostIt: (state, action: PayloadAction<IPostIt>) => {
			state.postItList.push({ ...action.payload, disabled: true });
			if (state.orderBy === 'DES') {
				sortDes(state.postItList);
			} else {
				sortAsc(state.postItList);
			}
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		restoreAllPostIts: (state, action: PayloadAction<Array<IPostIt>>) => {
			const restoredPostIts = action.payload.map((postIt) => ({
				...postIt,
				disabled: true,
			}));
			state.postItList = [...state.postItList, ...restoredPostIts];
			if (state.orderBy === 'DES') {
				sortDes(state.postItList);
			} else {
				sortAsc(state.postItList);
			}
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		moveToTrash: (state, action: PayloadAction<PostItId>) => {
			const postItIndex = state.postItList.findIndex(
				(postIt) => postIt.id === action.payload.id
			);
			state.postItList.splice(postItIndex, 1);
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		moveAllToTrash: (state) => {
			state.postItList = [];
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		editPostIt: (state, action: PayloadAction<IPostIt>) => {
			state.postItList.forEach((postIt) => {
				if (postIt.id === action.payload.id) {
					postIt.content = action.payload.content;
				}
			});
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		orderByDate: (state) => {
			if (state.orderBy === 'DES') {
				sortAsc(state.postItList);
				state.orderBy = 'ASC';
			} else {
				sortDes(state.postItList);
				state.orderBy = 'DES';
			}
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
			localStorage.setItem('orderBy', JSON.stringify(state.orderBy));
		},
		switchToDarkMode: (state) => {
			const modeStatus = document.documentElement.classList.toggle('dark');
			state.darkMode = modeStatus;
			localStorage.setItem('darkMode', JSON.stringify(state.darkMode));
		},
	},
});

export const {
	createPostIt,
	restorePostIt,
	restoreAllPostIts,
	moveToTrash,
	moveAllToTrash,
	editPostIt,
	orderByDate,
	switchToDarkMode,
} = workspaceSlice.actions;

export default workspaceSlice.reducer;
