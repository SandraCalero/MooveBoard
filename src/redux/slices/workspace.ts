/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface PostItProps {
	id: number;
	variant: string;
	content?: string;
	draggable?: boolean;
	disabled?: boolean;
}

interface workspaceStateProps {
	currentPostItId: number;
	postItList: Array<PostItProps>;
}

const local = localStorage.getItem('postItList');
const postItList: Array<PostItProps> = local ? JSON.parse(local) : [];

const initialState: workspaceStateProps = {
	currentPostItId: 0,
	postItList,
};

const workspaceSlice = createSlice({
	name: 'workspace',
	initialState,
	reducers: {
		createPostIt: (state) => {
			state.currentPostItId += 1;
			state.postItList.push({
				id: state.currentPostItId,
				content: '',
				variant: 'postItCreated',
				draggable: true,
			});
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		sendToTrash: (state, action) => {
			console.log('Send to trash', state, action.payload);
		},
		editPostIt(state, action: PayloadAction<PostItProps>) {
			state.postItList.forEach((postIt) => {
				if (postIt.id === action.payload.id) {
					postIt.content = action.payload.content;
				}
			});
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
	},
});

export const { createPostIt, sendToTrash, editPostIt } = workspaceSlice.actions;

export default workspaceSlice.reducer;
