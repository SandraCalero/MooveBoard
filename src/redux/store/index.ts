import { configureStore } from '@reduxjs/toolkit';
import workspaceReducer from 'redux/reducers/workspace';
import trashspaceReducer from 'redux/reducers/trashspace';

const store = configureStore({
	reducer: {
		workspace: workspaceReducer,
		trashspace: trashspaceReducer,
	},
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
