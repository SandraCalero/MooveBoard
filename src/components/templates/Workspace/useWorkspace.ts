import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { addPostItToTrash, addAllToTrash } from 'redux/reducers/trashspace';
import {
	createPostIt,
	moveAllToTrash,
	moveToTrash,
	orderByDate,
	switchToDarkMode,
} from 'redux/reducers/workspace';
import { RootState } from 'redux/store';
import { IPostIt } from 'globals/definitions/postItProps';
import orderASC from 'assets/icons/orderASC.png';
import orderDES from 'assets/icons/orderDES.png';

/**
 * It returns an object with all the functions and variables that we need to use in our workspace
 * component
 * @returns An object with the following properties:
 * isModalOpen,
 * postItList,
 * modeText,
 * showModal,
 * handleCreatePostIt,
 * movePostItToTrash,
 * handleMoveAllToTrash,
 * handleOrderByDate,
 * hideModal,
 * handleDarkMode,
 */
const useWorkspace = () => {
	const dispatch = useDispatch();

	/* Selectors */
	const postItList = useSelector(
		(state: RootState) => state.workspace.postItList
	);
	const darkMode = useSelector((state: RootState) => state.workspace.darkMode);
	const orderState = useSelector((state: RootState) => state.workspace.orderBy);

	/* Variables */
	const initialPostIt: IPostIt = { id: 0, content: '' };
	const modeText = darkMode ? 'Light' : 'Dark';
	const textOrderBy =
		orderState === 'DES' ? 'Order by oldest' : 'Order by newest';
	const iconOrderBy = orderState === 'DES' ? orderASC : orderDES;

	/* Initial states which will be used to handle the modal behavior */
	const [postIt, setPostIt] = useState(initialPostIt);
	const [isModalOpen, setIsModalOpen] = useState(false);

	/**
	 * It takes a postItInfo object as an argument, sets the isModalOpen state to true, and sets the postIt
	 * state to the postItInfo object
	 * @param {IPostIt} postItInfo - This is the post-it that was clicked on.
	 */
	const showModal = (postItInfo: IPostIt) => {
		setIsModalOpen(true);
		setPostIt(postItInfo);
	};

	/**
	 * It sets the state of the modal to false and resets the post-it to its initial state
	 */
	const hideModal = () => {
		setIsModalOpen(false);
		setPostIt(initialPostIt);
	};

	/**
	 * It dispatches two actions, one to move the post-it to the trash and one to add the post-it to the
	 * trash
	 */
	const movePostItToTrash = () => {
		dispatch(moveToTrash(postIt));
		dispatch(addPostItToTrash(postIt));
		hideModal();
	};

	/**
	 * It dispatches an action to the Redux store to create a new post it
	 */
	const handleCreatePostIt = () => {
		dispatch(createPostIt());
	};

	/**
	 * It dispatches the moveAllToTrash action, which sets the postItList to an empty array, and then
	 * dispatches the addAllToTrash action, which adds the postItList to the trashList
	 */
	const handleMoveAllToTrash = () => {
		dispatch(moveAllToTrash());
		dispatch(addAllToTrash(postItList));
	};

	/**
	 * It dispatches an action to the store that will sort the posts by date
	 */
	const handleOrderByDate = () => {
		dispatch(orderByDate());
	};

	/**
	 * It dispatches an action to the reducer to switch to dark mode
	 */
	const handleDarkMode = () => {
		dispatch(switchToDarkMode());
	};

	return {
		isModalOpen,
		postItList,
		modeText,
		iconOrderBy,
		textOrderBy,
		showModal,
		handleCreatePostIt,
		movePostItToTrash,
		handleMoveAllToTrash,
		handleOrderByDate,
		hideModal,
		handleDarkMode,
	};
};

export default useWorkspace;
