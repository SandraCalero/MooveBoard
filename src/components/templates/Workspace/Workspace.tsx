import Button from 'components/atoms/Button/Button';
import TrashBinLink from 'components/molecules/TrashBinLink/TrashBinLink';
import Header from 'components/organisms/Header/Header';
import newPostIt from 'assets/icons/newPostIt.png';
import sendAllToTrash from 'assets/icons/sendAllToTrash.png';
import PostItEditableList from 'components/organisms/PostItEditableList/PostItEditableList';
import Modal from 'components/molecules/Modal/Modal';
import useWorkspace from './useWorkspace';

export default function Workspace() {
	const {
		isModalOpen,
		postItList,
		modeText,
		movePostItToTrash,
		hideModal,
		showModal,
		handleCreatePostIt,
		handleMoveAllToTrash,
		handleDarkMode,
	} = useWorkspace();

	return (
		<main className='overflow-y-scroll h-full flex flex-col'>
			<Header>
				<Button
					variant='w-20 h-20'
					icon={newPostIt}
					altText='Create new Post It'
					onClick={handleCreatePostIt}
				/>
				<Button
					variant='w-20 h-20 '
					text='Delete All'
					icon={sendAllToTrash}
					altText='Send All to Trash Bin'
					onClick={handleMoveAllToTrash}
				/>
				<TrashBinLink />
			</Header>
			<section>
				<h1
					className='m-2 text-center text-grey dark:text-mine-shaft-400
				transition-colors'
				>
					Workspace
				</h1>
				<PostItEditableList postItList={postItList} showModal={showModal} />
				<Modal
					title='Move post it to the trash'
					message='Are you sure you want to move this post it to the trash?'
					isModalOpen={isModalOpen}
					onConfirm={movePostItToTrash}
					onCancel={hideModal}
				/>
			</section>

			<Button
				variant='w-20 fixed bottom-2 right-4 bg-blue text-white p-2
				hover:opacity-100 font-medium after:absolute after:right-1
				after:bg-white !items-start	after:rounded-full after:p-1
				after:w-8 after:h-8 after:text-sm active:translate-y-0
				dark:!items-end dark:after:left-1 dark:after:bg-blue
				dark:bg-white dark:text-blue rounded-3xl transition-colors
				transition-transform duration-300 md:w-24'
				text={modeText}
				onClick={handleDarkMode}
			/>
		</main>
	);
}
