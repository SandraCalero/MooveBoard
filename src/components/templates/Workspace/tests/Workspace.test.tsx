import { BrowserRouter } from 'react-router-dom';
import { createPostIt, editPostIt } from 'redux/reducers/workspace';
import store from 'redux/store';
import { render } from 'utils/tests/test-utils';
import Workspace from '../Workspace';

describe('<Workspace />', () => {
	it('should match the snapshot', () => {
		store.dispatch(createPostIt());
		store.dispatch(
			editPostIt({ id: 1, content: 'Testing edit post it action' })
		);
		const { container, getByText } = render(
			<BrowserRouter>
				<Workspace />
			</BrowserRouter>
		);
		expect(container).toMatchSnapshot();
		expect(getByText('Testing edit post it action')).toBeTruthy();
	});
});
