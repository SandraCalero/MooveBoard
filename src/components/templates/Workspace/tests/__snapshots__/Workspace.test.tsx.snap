// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Workspace /> should match the snapshot 1`] = `
<div>
  <main
    class="overflow-y-scroll h-full flex flex-col"
  >
    <header
      class="flex justify-around items-center p-3"
    >
      <button
        class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 w-20 h-20"
        name="button"
        type="button"
      >
        <img
          alt="Create new Post It"
          class="h-4/5 w-4/5"
          src="newPostIt.png"
        />
      </button>
      <button
        class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 w-20 h-20 "
        name="button"
        type="button"
      >
        <img
          alt="Send All to Trash Bin"
          class="h-4/5 w-4/5"
          src="sendAllToTrash.png"
        />
        Delete All
      </button>
      <a
        class="flex flex-col items-center no-underline p-0 m-2 text-center hover:opacity-90 active:translate-y-1"
        href="/trashbin"
      >
        <img
          alt="Trash bin"
          class="w-16 h-16"
          src="trashBinEmpty.png"
        />
        <span>
          Trash Bin
        </span>
      </a>
    </header>
    <section>
      <h1
        class="m-2 text-center text-grey dark:text-mine-shaft-400 transition-colors"
      >
        Workspace
      </h1>
      <div
        class="px-4 flex justify-start items-center"
      >
        <button
          class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 h-10 w-10"
          name="button"
          type="button"
        >
          <img
            alt=""
            class="h-4/5 w-4/5"
            src="orderASC.png"
          />
        </button>
        <span
          class="text-sm m-0 p-0"
        >
          Order by oldest
        </span>
      </div>
      <ul
        class="grid grid-cols-auto p-4 m-0 list-none gap-4 justify-center"
      >
        <li
          class="justify-self-center"
        >
          <div
            aria-hidden="true"
            class="flex flex-col items-center justify-center bg-yellow shadow-xl w-max dark:bg-cerulean-500 dark:shadow-lg dark:shadow-cerulean-700 transition-colors dark:border dark:border-grey dark:text-mine-shaft-600"
            id="1"
          >
            <button
              class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 my-0 mx-0 px-2 pt-2 opacity-20 text-xl hover:opacity-100 self-end dark:text-mine-shaft-900 dark:opacity-40 dark:hover:opacity-100 transition-colors"
              name="button"
              type="button"
            >
              X
            </button>
            <textarea
              class="p-2 w-52 h-52 bg-transparent outline-none border-none resize-y focus:cursor-text"
              disabled=""
              name="note"
            >
              Testing edit post it action
            </textarea>
          </div>
        </li>
      </ul>
      <div
        class="hidden items-center justify-center z-10 m-0 p-0 w-full
			h-full absolute inset-0 bg-transparent"
      >
        <div
          class="flex flex-col justify-center items-center bg-white max-w-xs h-auto rounded-lg shadow-2xl shadow-mine-shaft-900 dark:shadow-cerulean-700 dark:bg-cerulean-900 dark:border-none transition-colors"
        >
          <h2
            class="flex items-center p-3 m-2"
          >
            Move post it to the trash
          </h2>
          <div
            class="flex items-center p-3 m-0 border-y border-solid border-grey"
          >
            <img
              alt="Warning icon"
              class="m-2 w-10"
              src="warning.png"
            />
            <p
              class="m-2 text-sm "
            >
              Are you sure you want to move this post it to the trash?
            </p>
          </div>
          <div
            class="flex items-center p-3 m-0"
          >
            <button
              class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 justify-center w-24 h-12 rounded-lg bg-grey dark:bg-mine-shaft-500 dark:text-white transition-colors"
              name="button"
              type="button"
            >
              Cancel
            </button>
            <button
              class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 justify-center w-24 h-12 rounded-lg bg-blue text-white dark:text-blue dark:bg-white font-semibold transition-colors"
              name="button"
              type="button"
            >
              Confirm
            </button>
          </div>
        </div>
      </div>
    </section>
    <button
      class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 w-20 fixed bottom-2 right-4 bg-blue text-white p-2 hover:opacity-100 font-medium after:absolute after:right-1 after:bg-white !items-start	after:rounded-full after:p-1 after:w-8 after:h-8 after:text-sm active:translate-y-0 dark:!items-end dark:after:left-1 dark:after:bg-blue dark:bg-white dark:text-blue rounded-3xl transition-colors transition-transform duration-300 md:w-24"
      name="button"
      type="button"
    >
      Dark
    </button>
  </main>
</div>
`;
